patches/server/1051-Add-FluidState-API.patch

```diff
@@ -0,0 +1,242 @@
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/block/fluid/PaperFluidData.java
+@@ -0,0 +1,132 @@
++package io.papermc.paper.block.fluid;
++
++import com.google.common.base.Preconditions;
++import io.papermc.paper.block.fluid.type.PaperFlowing;
++import io.papermc.paper.block.fluid.type.PaperFallingFluid;
++import java.util.HashMap;
++import java.util.Map;
++import java.util.function.Function;
++import net.minecraft.core.BlockPos;
++import net.minecraft.world.level.block.state.properties.Property;
++import net.minecraft.world.level.material.FluidState;
++import net.minecraft.world.level.material.LavaFluid;
++import net.minecraft.world.level.material.WaterFluid;
++import org.bukkit.Fluid;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.craftbukkit.CraftFluid;
++import org.bukkit.craftbukkit.CraftWorld;
++import org.bukkit.craftbukkit.util.CraftVector;
++import org.bukkit.util.Vector;
++
++public class PaperFluidData implements FluidData {
++
++    private static final Map<Class<? extends net.minecraft.world.level.material.Fluid>, Function<FluidState, PaperFluidData>> MAP = new HashMap<>();
++
++    private final FluidState state;
++
++    protected PaperFluidData(final FluidState state) {
++        this.state = state;
++    }
++
++    /* Registry */
++
++    static {
++        //<editor-fold desc="PaperFluidData Registration" defaultstate="collapsed">
++        register(LavaFluid.Source.class, PaperFallingFluid::new);
++        register(WaterFluid.Source.class, PaperFallingFluid::new);
++        register(LavaFluid.Flowing.class, PaperFlowing::new);
++        register(WaterFluid.Flowing.class, PaperFlowing::new);
++        //</editor-fold>
++    }
++
++    static void register(final Class<? extends net.minecraft.world.level.material.Fluid> fluid, final Function<FluidState, PaperFluidData> creator) {
++        Preconditions.checkState(MAP.put(fluid, creator) == null, "Duplicate mapping %s->%s", fluid, creator);
++        MAP.put(fluid, creator);
++    }
++
++    public static PaperFluidData createData(final FluidState state) {
++        return MAP.getOrDefault(state.getType().getClass(), PaperFluidData::new).apply(state);
++    }
++
++    /* Impl */
++
++    public FluidState getState() {
++        return this.state;
++    }
++
++    protected <T extends Comparable<T>> T get(final Property<T> property) {
++        return this.state.getValue(property);
++    }
++    /* API */
++
++    @Override
++    public final Fluid getFluidType() {
++        return CraftFluid.minecraftToBukkit(this.state.getType());
++    }
++
++    @Override
++    public PaperFluidData clone() {
++        try {
++            return (PaperFluidData) super.clone();
++        } catch (CloneNotSupportedException ex) {
++            throw new AssertionError("Clone not supported", ex);
++        }
++    }
++
++    @Override
++    public Vector getFlowDirection(World world, Location location) {
```

Comment:
Remove the separate world argument, because a location contains a world

Reply:
Location can not contain a World. Which would mean i would have to return null or make it throw an exception

Reply:
You could add an not null check using `Preconditions.checkNotNull(location.getWorld())` or something similar, because calling this method like this seems a bit silly:
```java
fluid.getFlowDirection(location.getWorld(), location);
```